src/outputs/fifo.c

```diff
@@ -0,0 +1,498 @@
+  /* Do not dereference - only passed to the status cb */
+  struct output_device *device;
+  struct output_session *output_session;
+  output_status_cb status_cb;
+};
+
+/* From player.c */
+extern struct event_base *evbase_player;
+
+static struct fifo_session *sessions;
+
+/* Forwards */
+static void
+defer_cb(int fd, short what, void *arg);
+
+
+/* ---------------------------- FIFO HANDLING ---------------------------- */
+
+static void
+fifo_delete(struct fifo_session *fifo_session)
+{
+  DPRINTF(E_DBG, L_FIFO, "Removing FIFO \"%s\"\n", fifo_session->path);
+
+  if (unlink(fifo_session->path) < 0)
+    {
+      DPRINTF(E_WARN, L_FIFO, "Could not remove FIFO \"%s\": %d\n", fifo_session->path, errno);
+      return;
+    }
+
+  fifo_session->created = 0;
+}
+
+static void
+fifo_close(struct fifo_session *fifo_session)
+{
+  struct stat st;
+
+  if (fifo_session->input_fd >= 0)
+    {
+      close(fifo_session->input_fd);
+      fifo_session->input_fd = -1;
+    }
+
+  if (fifo_session->output_fd >= 0)
+    {
+      close(fifo_session->output_fd);
+      fifo_session->output_fd = -1;
+    }
+
+  if (fifo_session->created && (stat(fifo_session->path, &st) == 0))
+    fifo_delete(fifo_session);
+}
+
+static int
+fifo_make(struct fifo_session *fifo_session)
+{
+  if (mkfifo(fifo_session->path, 0666) < 0)
+    {
+      DPRINTF(E_LOG, L_FIFO, "Could not create FIFO \"%s\": %d\n", fifo_session->path, errno);
+      return -1;
+    }
+
+  fifo_session->created = 1;
+
+  return 0;
+}
+
+static int
+fifo_check(struct fifo_session *fifo_session)
+{
+  struct stat st;
+
+  if (stat(fifo_session->path, &st) < 0)
+    {
+      if (errno == ENOENT)
+	{
+	  /* Path doesn't exist */
+	  return fifo_make(fifo_session);
+	}
+
+      DPRINTF(E_LOG, L_FIFO, "Failed to stat FIFO \"%s\": %d\n", fifo_session->path, errno);
+      return -1;
+    }
+
+  if (!S_ISFIFO(st.st_mode))
+    {
+      DPRINTF(E_LOG, L_FIFO, "\"%s\" already exists, but is not a FIFO\n", fifo_session->path);
+      return -1;
+    }
+
+  return 0;
+}
+
+static int
+fifo_open(struct fifo_session *fifo_session)
+{
+  int ret;
+
+  ret = fifo_check(fifo_session);
+  if (ret < 0)
+    return -1;
+
+  fifo_session->input_fd = open(fifo_session->path, O_RDONLY | O_NONBLOCK, 0);
+  if (fifo_session->input_fd < 0)
+    {
+      DPRINTF(E_LOG, L_FIFO, "Could not open FIFO \"%s\" for reading: %d\n", fifo_session->path, errno);
+      fifo_close(fifo_session);
+      return -1;
+    }
+
+  fifo_session->output_fd = open(fifo_session->path, O_WRONLY | O_NONBLOCK, 0);
+  if (fifo_session->output_fd < 0)
+    {
+      DPRINTF(E_LOG, L_FIFO, "Could not open FIFO \"%s\" for writing: %d\n", fifo_session->path, errno);
+      fifo_close(fifo_session);
+      return -1;
+    }
+
+  return 0;
+}
+
+static void
+fifo_empty(struct fifo_session *fifo_session)
+{
+  char buf[FIFO_BUFFER_SIZE];
+  int bytes = 1;
+
+  while (bytes > 0 && errno != EINTR)
+    bytes = read(fifo_session->input_fd, buf, FIFO_BUFFER_SIZE);
```

Comment:
flush can be called without the fifo having been created, which can lead to a crash. Here is the output from Valgrind:

==16132== Invalid read of size 4
==16132==    at 0x808B65D: read (unistd.h:44)
==16132==    by 0x808B65D: fifo_empty.isra.0 (fifo.c:185)
==16132==    by 0x808B6E9: fifo_flush (fifo.c:386)
==16132==    by 0x80841F6: outputs_flush (outputs.c:194)
==16132==    by 0x807FB9E: playback_stop (player.c:2215)
==16132==    by 0x80973A7: command_cb_sync (commands.c:84)
==16132==    by 0x80973A7: command_cb (commands.c:141)
==16132==    by 0x5913233: event_base_loop (in /usr/lib/i386-linux-gnu/libevent-2.0.so.5.1.9)
==16132==    by 0x5913FE2: event_base_dispatch (in /usr/lib/i386-linux-gnu/libevent-2.0.so.5.1.9)
==16132==    by 0x807F2BD: player (player.c:3938)
==16132==    by 0x5C5EF71: start_thread (pthread_create.c:312)
==16132==    by 0x5D5FF8D: clone (clone.S:129)
==16132==  Address 0x8 is not stack'd, malloc'd or (recently) free'd
==16132== 
==16132== 
==16132== Process terminating with default action of signal 11 (SIGSEGV)
==16132==  Access not within mapped region at address 0x8
==16132==    at 0x808B65D: read (unistd.h:44)
==16132==    by 0x808B65D: fifo_empty.isra.0 (fifo.c:185)
==16132==    by 0x808B6E9: fifo_flush (fifo.c:386)
==16132==    by 0x80841F6: outputs_flush (outputs.c:194)
==16132==    by 0x807FB9E: playback_stop (player.c:2215)
==16132==    by 0x80973A7: command_cb_sync (commands.c:84)
==16132==    by 0x80973A7: command_cb (commands.c:141)
==16132==    by 0x5913233: event_base_loop (in /usr/lib/i386-linux-gnu/libevent-2.0.so.5.1.9)
==16132==    by 0x5913FE2: event_base_dispatch (in /usr/lib/i386-linux-gnu/libevent-2.0.so.5.1.9)
==16132==    by 0x807F2BD: player (player.c:3938)
==16132==    by 0x5C5EF71: start_thread (pthread_create.c:312)
==16132==    by 0x5D5FF8D: clone (clone.S:129)